components/api/src/error.rs

```diff
@@ -14,49 +14,116 @@
-use serde::Serialize;
-use std::io;
-
-/// List of possible API errors.
-#[derive(Fail, Debug)]
-pub enum Error {
-    /// Storage error. This type includes errors related to the database, caused
-    /// by, for example, serialization issues.
-    #[fail(display = "Storage error: {}", _0)]
-    Storage(#[cause] failure::Error),
-
-    /// Input/output error. This type includes errors related to files that are not
-    /// a part of the Exonum storage.
-    #[fail(display = "IO error: {}", _0)]
-    Io(#[cause] io::Error),
-
-    /// Bad request. This error occurs when the request contains invalid syntax.
-    #[fail(display = "Bad request: {}", _0)]
-    BadRequest(String),
-
-    /// Moved permanently. This error means that resource existed at the specified
-    /// location, but now is moved to the other place.
-    #[fail(display = "Moved permanently; Location: {}", _0)]
-    MovedPermanently(String),
-
-    /// Gone. This error means that resource existed in the past, but now is not present.
-    #[fail(display = "Gone")]
-    Gone,
-
-    /// Not found. This error occurs when the server cannot locate the requested
-    /// resource.
-    #[fail(display = "Not found: {}", _0)]
-    NotFound(String),
-
-    /// Internal server error. This type can return any internal server error to the user.
-    #[fail(display = "Internal server error: {}", _0)]
-    InternalError(failure::Error),
-
-    /// Unauthorized error. This error occurs when the request lacks valid
-    /// authentication credentials.
-    #[fail(display = "Unauthorized")]
-    Unauthorized,
+use serde::{Deserialize, Serialize};
+
+use std::fmt;
+
+/// API HTTP error struct.
+#[derive(Fail, Debug, PartialEq)]
+pub struct Error {
+    /// HTTP error code.
+    pub http_code: HttpStatusCode,
+    /// API error body.
+    pub body: ErrorBody,
+    /// Additional HTTP headers.
+    pub headers: HeaderMap,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default, PartialEq)]
+pub struct ErrorBody {
+    /// A URI reference to the documentation or possible solutions for the problem.
+    #[serde(rename = "type", default, skip_serializing_if = "String::is_empty")]
+    pub docs_uri: String,
+    /// Short description of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub title: String,
+    /// Detailed description of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub detail: String,
+    /// Source of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub source: String,
+    /// Internal error code.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub error_code: Option<u8>,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: {}", self.body.title, self.body.detail)
+    }
+}
+
+impl Error {
+    /// Builds a ApiError with the given `http_code`.
+    pub fn new(http_code: HttpStatusCode) -> Self {
```

Comment:
Could you add constructors for common erroneous status codes, e.g. `bad_request` (400) / `not_found` (404) / `forbidden` (403) / `internal` (500)? Also, the internal constructor may accept the error to immediately set the `detail`:

```rust
pub fn internal(cause: impl failure::Fail) { /* ... */ }
```